AlphaWallet/Settings/ViewControllers/ExportJsonKeystorePasswordViewController.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  Created by Jerome Chan on 1/12/21.
+//
+
+import UIKit
+
+@objc protocol ExportJsonKeystorePasswordDelegate {
+    func didRequestExportKeystore(with password: String)
+    func didDismissPasswordController()
+}
+
+class ExportJsonKeystorePasswordViewController: UIViewController {
+    private let buttonTitle: String
+    private var viewModel: ExportJsonKeystorePasswordViewModel
+    private lazy var keyboardChecker = KeyboardChecker(self, resetHeightDefaultValue: 0, ignoreBottomSafeArea: true)
+    private lazy var queue: DispatchQueue = DispatchQueue(label: "ExportJsonKeystorePasswordViewController")
+    private var passwordView: ExportJsonKeystorePasswordView {
+        return view as! ExportJsonKeystorePasswordView
+    }
+    weak var passwordDelegate: ExportJsonKeystorePasswordDelegate?
+
+    init(viewModel: ExportJsonKeystorePasswordViewModel, buttonTitle: String) {
+        self.viewModel = viewModel
+        self.buttonTitle = buttonTitle
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        return nil
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureController()
+        passwordView.passwordTextField.becomeFirstResponder()
+        passwordView.disableButton()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        keyboardChecker.viewWillAppear()
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        keyboardChecker.viewWillDisappear()
+    }
+
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        if !isStillInNavigationStack() {
+            passwordDelegate?.didDismissPasswordController()
+        }
+    }
+
+    override func loadView() {
+        view = ExportJsonKeystorePasswordView()
+    }
+
+    private func configureController() {
+        navigationItem.title = R.string.localizable.settingsAdvancedExportJSONKeystorePasswordTitle()
+        passwordView.setButton(title: buttonTitle)
+        passwordView.addExportButtonTarget(self, action: #selector(requestExportAction(_:)))
+        passwordView.passwordTextField.delegate = self
+        keyboardChecker.constraint = passwordView.bottomConstraint
+    }
+
+    @objc func requestExportAction(_ sender: UIButton?) {
+        guard let password = passwordView.passwordTextField.text else { return }
+        switch viewModel.validate(password: password) {
+        case .success:
+            computeExportJsonData(password: password)
+        case .failure:
+            break
+        }
+    }
+
+    private func computeExportJsonData(password: String) {
+        view.endEditing(true)
+        navigationController?.displayLoading(text: "Exporting", animated: true)
+        DispatchQueue.main.async {
+            self.queue.async {
+                self.viewModel.computeHash(password: password) { result in
```

Comment:
The func `computeHash(password:)`'s name doesn't look intention-revealing

Reply:
I think it is slightly better to just let this view controller capture the password, validate it and then pass it on to the next screen which generates the JSON result, rather than generate the JSON and pass it around to reduce places where we capture sensitive data.

BTW, I presume you tested that the async/sync flow works properly with Face ID/Touch ID? Because this line would trigger it:

`keystore.exportRawPrivateKeyForNonHdWalletForBackup(`

and Face ID and Touch ID can be sensitive to which thread you call them from.